### PR TITLE
add missing C-linkage in headers

### LINKED ISSUE #77605: Some headers in include/zephyr/drivers/gpio are missing the C++ compatibility
**Describe the bug**
The following headers in include/zephyr/drivers/gpio are missing the extern "C":

- gpio_intel.h (no functions there, probably not necessary anyway)
- gpio_mmio32.h
- gpio_nct38xx.h
- gpio_sx1509b.h
- gpio_utils.h

### PR DESCRIPTION
**Pull request by benediktibk**
Add missing extern "C" statements in headers for GPIO drivers. Fixes #77605.

### CODE REVIEW

**Review by bjarki-andreasen — DISMISSED** [review]
The changes to gpio_utils.h may not be necessary (don't know if this file is used outside of the drivers themselves, even though it is placed in a public header), but for completeness its fine :)

File: include/zephyr/drivers/gpio/gpio_intel.h
Comment (henrikbrixandersen) [inline include/zephyr/drivers/gpio/gpio_intel.h:1]: This one is missing an include guard as well.

**Review by henrikbrixandersen — DISMISSED** [review]

File: include/zephyr/drivers/gpio/gpio_intel.h
Comment (benediktibk) [reply include/zephyr/drivers/gpio/gpio_intel.h:1]: I just realized soc/intel/common/soc_gpio.h as well. And I cannot find any include for gpio_intel.h, although the struct defined in it is in use, although within an #if. Smells a little bit like rotten code.

**Comment by benediktibk** [comment]
I assume the build error is not caused by this PR.

    INFO    -  27/664 m2gl025_miv               tests/drivers/gnss/gnss_nmea0183/drivers.gnss.gnss_nmea0183  FAILED Timeout (renode 120.883s)
    INFO    - /__w/zephyr/zephyr/twister-out/m2gl025_miv/tests/drivers/gnss/gnss_nmea0183/drivers.gnss.gnss_nmea0183/handler.log
    ERROR   - [0/1] cd /__w/zephyr/zephyr/twister-out/m2gl025_miv/tests/drivers/gnss/gnss_nmea0183/drivers.gnss.gnss_nmea0183 && /usr/bin/renode --disable-xwt --port -2 --pid-file renode.pid -e '$elf=@/__w/zephyr/zephyr/twister-out/m2gl025_miv/tests/drivers/gnss/gnss_nmea0183/drivers.gnss.gnss_nmea0183/zephyr/zephyr.elf; include @/__w/zephyr/zephyr/boards/microchip/m2gl025_miv/support/m2gl025_miv.resc; s'
    09:18:28.0950 [INFO] Loaded monitor commands from: /opt/renode/scripts/monitor.py
    
    INFO    - /__w/zephyr/zephyr/twister-out/m2gl025_miv/tests/drivers/gnss/gnss_nmea0183/drivers.gnss.gnss_nmea0183/handler.log

**Review by henrikbrixandersen — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Pull request merged**